### DISCUSSION TITLE
[RFC] Find popper.js alternative to create accessible Storybook UI dropdown components

### DISCUSSION THREAD

**Discussion by danielmarcano** [storybookjs/storybook#22415] [RFC]
**Status: accepted; championed by @valentinpalkovic**

-----

### Summary

Finding an alternative to [popper.js](https://popper.js.org/) for re-building all Storybook UI dropdowns, so that they are more accessible and keyword-navigatable.

Ideally, this would be a framework-agnostic UI components library, such as [Shoelace](https://shoelace.style/components/menu), or a React-only library, like [Radix UI](https://www.radix-ui.com/) or [React Aria](https://react-spectrum.adobe.com/react-aria/useMenu.html).

Credits to @MichaelArestad and @JReinhold for their suggestions.

### Problem Statement

Currently, Storybook's dropdowns are not accessible via keyboard navigation, which can create barriers for users who rely on their keyboards in order to interact with / navigate through software.

As Storybook keeps growing, accessibility—a fundamental aspect in software—becomes a stronger necessity, so that everyone can have a great User Experience and take the most advantage out of the library.

All things said, there's two ways to approach this:

- Building fully custom components (and keeping `popper.js`)
- Using a UI component library, such as ([React Aria](https://react-spectrum.adobe.com/react-aria/useMenu.html)

The disadvantage of the first approach being that `popper.js` is meant to be used for building tooltips, instead of dropdown menus, which means that enhancing the user-experience of the existing dropdowns requires a lot of manual work: handling keyboard events properly, taking care of any edge-cases, and being compliant with the [W3C's Accessibility Guidelines](https://w3c.github.io/silver/guidelines/).

The advantage of the latter one, which is the proposed solution, is that these components will not need to be maintained—from an accessibility point of view—by the Storybook team (nor the community), so we can simply choose the UI component library that suits us best: whether a framework-agnostic one, or a React-only one.

### Non-goals

- Using the UI component library we choose for any other Storybook UI components which are not menu dropdowns
- Redesigning Storybook UI menu dropdown components
- Modifying the current behavior of all existing menu dropdown components

### Implementation

TBD

### Deliverables

1. Decide on a UI component library
2. Import the library's packages
3. Replace existing dropdowns with the library's menu dropdown components
4. Adapt the UI component library components styles to Storybook dropdown styles
5. Make sure that the dropdowns' functionality remains the same, even after being re-built, since this update should only make the components more accessible, no more, no less
6. Ensure that no tests are broken after these changes
7. Create a new PR which delivers this enhancement (thus closing [this issue](https://github.com/storybookjs/storybook/issues/21578))

### Risks

- This enhancement adds a new dependency (or a couple of them) to the project, which increases its bundling size, and also requires for us to keep it up to date.

### Unresolved Questions

- [ ] Will the UI component library be compatible with all current Storybook components?
- [ ] Will the customizations needed to fit the chosen UI component library into the Storybook UI be extensive?
- [ ] Will any additional training or documentation be needed for developers to use the new component library effectively?

### Alternatives considered / Abandoned Ideas

- Adding the necessary accessibility enhancements manually, to the existing `popper.js` dropdowns. More info on [this closed issue](https://github.com/storybookjs/storybook/pull/12904) from 2020.

- Using [Reach UI](https://reach.tech/). As mentioned by @valentinpalkovic, the library is no longer being maintained. It also does not support React 18, and Storybook is on its way to support it.

**Comment by valentinpalkovic** [comment]
Thank you very much for your RFC. Much appreciated!

Currently, I see the following (potential) issues with introducing Reach UI:

- It seems to be unmaintained: https://github.com/reach/reach-ui/issues/972
- React 18 isn't supported: https://github.com/reach/reach-ui/issues/984. We will move the Storybook manager soon-ish to React v18, though: https://github.com/storybookjs/storybook/pull/21673 
- Although the manager is written in React currently, that might change in the future. I have concerns about complicating potential transformation steps by adding a further react-specific third-party dependency.

I think the main issue currently is that we are not using popper.js in an accessible way. We should spawn the menu elements next to the button/element which triggers the menu opening. Currently, that's not the case, and the elements are placed at the `document.body` directly. Changing this behavior would enable us to use proper keyboard navigation and make the menu overlays accessible.

**Reply by danielmarcano** [reply]
Hi, @valentinpalkovic, thank you for the information!

I will actually go ahead and add those three risks to the RFC. This is my first time contributing to such a large Open Source project, so there's a couple of things that I still need to improve upon, one of them is definitely checking if any external libraries proposed are being actively maintained.

Those three reasons, though, seem like a dealbreaker to me, or in other words, lead me to think that Reach UI is no longer the desired library for this enhancement.

That takes us to [Radix UI](https://www.radix-ui.com/), one of the alternatives I checked, which was suggested by @MichaelArestad.

That is actually one of the alternatives that is suggested in the first issue you have shared.

As for popper.js, I honestly don't think that's the right tool for this job, since the library is aimed solely at creating tooltips, and Storybook dropdowns fall in the "menu button" category. Trying to use the library to solve this issue would require a similar amount of manual work as though we were creating a fully custom component from scratch, which might not be what we want, according to [this closed issue](https://github.com/storybookjs/storybook/pull/12904) from 2020 and some of the reasons stated in this RFC.

Last but not least: is the React "detachment" expected to happen any time soon?

From an subjective point of view, adding a new React dependency does not seem like a constraint, specially since, given that moment, the entire UI components would have to be "moved away" from React code.

What are your thoughts on this?

Also: what do you recommend me do with this RFC, given the fact that Reach UI would actually not work as it is not maintained, and it's also not compatible with React 18?

And thanks again for the support and for the insight ✨

**Reply by valentinpalkovic** [reply]
> Last but not least: is the React "detachment" expected to happen soon?

It is not planned yet but might happen sooner or later. 

> As for popper.js, I honestly don't think that's the right tool for this job, since the library is aimed solely at creating tooltips, and Storybook dropdowns fall in the "menu button" category. Trying to use the library to solve this issue would require a similar amount of manual work as though we were creating a fully custom component from scratch, which might not be what we want, according to https://github.com/storybookjs/storybook/pull/12904 from 2020 and some of the reasons stated in this RFC.

I agree. Nevertheless, as popper.js is already in place, I would at least recommend investing some time to figure out, whether the accessibility issues are solvable with popper.js or not. I don't think, that all possibilities have been tried out so far.

> what do you recommend me do with this RFC, given the fact that Reach UI would actually not work as it is not maintained, and it's also not compatible with React 18?

You could either open a new RFC with a different UI library, or you can use this RFC and relabel it. We can try to figure out here, which particular UI library/solution we should take into account to finally decide on one. So maybe the goal of this RFC should be to find a good canditade/solution to replace the popper.js implementation.

-------------

I wonder whether we could also look at UI libraries, which are framework independent and rely on web components instead. I don't know whether there are some headless web component UI libraries out there.

**Reply by JReinhold** [reply]
Great RFC @danielmarcano! We have been talking about accessibility issues in the Storybook manager UI, and maybe this could be a first step of improving that.

> Last but not least: is the React "detachment" expected to happen any time soon?

It might happen at some point, which will be a big lift. I don't think that should stop us from relying on a React-only library, since we would need to rewrite the UI from scratch either way. Theoretically, if we did use a web components library for this, we could copy and paste it when migrating, but in reality I doubt this would be possible (state paradigms would be different, reactivity, etc.).

Other than Radix UI (which I think is a good solution), we could also explore [React Aria](https://react-spectrum.adobe.com/react-aria/useMenu.html) - those are the two I know of.
For a web components based solution, we could look at [Shoelace](https://shoelace.style/components/menu), which also has a [React wrapper](https://shoelace.style/frameworks/react), however it isn't really headless.

> Also: what do you recommend me do with this RFC, given the fact that Reach UI would actually not work as it is not maintained, and it's also not compatible with React 18?

Agree with @valentinpalkovic, feel free to re-purpose this RFC and just change titles and description (maybe with a historical side note), or create a new one.

**Comment by danielmarcano** [comment]
Thank you both, @valentinpalkovic and @JReinhold!

I have re-purposed the RFC based on your comments.

I will check the UI libraries mentioned here, and leave my findings in this RFC taking into account:

- Ease of use of the API
- How flexible they are in terms of re-styling
- Size of the packages
- Status / whether they are actively maintained
- React 18 support
- TypeScript support

Any other points you consider worth taking into account?

**Reply by JReinhold** [reply]
Looks great! The only addition to your list would be to gauge the breadth of the library, ie. how many other component types it contains, should we choose to expand the usage beyond menus at a later stage.

**Comment by danielmarcano** [comment]
Hi, there!

After carefully reviewing each of these libraries, testing them out, and checking the existing Storybook codebase and how the components have been crafted, I have reached the following conclusion: React Aria would fit best with the existing codebase, as it just provides the behavior and accessibility through React hooks that we can apply to any component, perhaps having to, at most, extend a couple of prop types for some of them, or creating a wrapper for avoiding this prop extension, for example.

Here's an overview of React Aria:

- The API is quite easy to use, although it requires some boilerplate, and the package has a "dependency" with react-stateful that is hard to break, as it would require either choosing another Design System state management library and adapt it to react-aria's types, or building this code from scratch

- It's 100% flexible, and fully agnostic to styling

- It's actively maintained, and has over 130K weekly downloads.

- The unpacked size of the [react-aria](https://www.npmjs.com/package/react-aria) library is 42.5KB. There's a catch, though. In order to use react-aria, the docs invite us to add [react-stately](https://www.npmjs.com/package/react-stately), whose unpacked size is 27.9 kB. According to npm, react-stately is "a library of React Hooks that provides cross-platform state management for your design system," and we can use it to create the state of react-aria's components. We can use other libraries for handling this, such as xstate, as mentioned in [this discussion](https://github.com/adobe/react-spectrum/discussions/2667), or even a custom hook, although in the same discussion one of the collaborators of react-aria does mention that creating these custom hooks manually might not have that many advantages

- It fully supports React 18, and have done so for over a year, as mentioned in these [release notes](https://react-spectrum.adobe.com/releases/2022-05-27.html)

- It fully supports TypeScript. Both [react-aria](https://www.npmjs.com/package/react-aria) and [react-stately](https://www.npmjs.com/package/react-stately) come with built-in TypeScript declarations

- These are some of the components it provides accessible behavior for:
  - [useMenu](https://react-spectrum.adobe.com/react-aria/useMenu.html)
  - [useTable](https://react-spectrum.adobe.com/react-aria/useTable.html)
  - [useCalendar](https://react-spectrum.adobe.com/react-aria/useCalendar.html)
  - [useCheckbox](https://react-spectrum.adobe.com/react-aria/useCheckbox.html)
  - [useCheckboxGroup](https://react-spectrum.adobe.com/react-aria/useCheckboxGroup.html)
  - [useRadioGroup](https://react-spectrum.adobe.com/react-aria/useRadioGroup.html)
  - [useSlider](https://react-spectrum.adobe.com/react-aria/useSlider.html)
  - [useSwitch](https://react-spectrum.adobe.com/react-aria/useSwitch.html)
  - [useTextField](https://react-spectrum.adobe.com/react-aria/useTextField.html)
  - [useBreadcrumbs](https://react-spectrum.adobe.com/react-aria/useBreadcrumbs.html)
  - [useLink](https://react-spectrum.adobe.com/react-aria/useLink.html)
  - [useTabList](https://react-spectrum.adobe.com/react-aria/useTabList.html)
  - [useDialog](https://react-spectrum.adobe.com/react-aria/useDialog.html)
  - [useModalOverlay](https://react-spectrum.adobe.com/react-aria/useModalOverlay.html)
  - [usePopover](https://react-spectrum.adobe.com/react-aria/usePopover.html)
  - [useSelect](https://react-spectrum.adobe.com/react-aria/useSelect.html)
  - [useProgressBar](https://react-spectrum.adobe.com/react-aria/useProgressBar.html)
  - [useToast](https://react-spectrum.adobe.com/react-aria/useToast.html)* beta

What are your thoughts on this?

**Comment by valentinpalkovic** [comment]
Hi @danielmarcano,

Thank you so much for wrapping this up. I wonder whether you could create a kind of comparison table to understand why react-aria was chosen and why, e.g., Radix UI wasn't. I am also wondering whether you took https://floating-ui.com/ into account since it is the accessible derivative of popper.js. Since we haven't decided yet, whether other components should be replaced by premade accessible ones as well, a quick win could be achieved by using floating-ui without introducing potential disruptive changes. I can't wait to hear your answer :)

**Reply by cdedreuille** [reply]
For reference, Radix UI uses floating-ui in the background for some of their components.

**Comment by danielmarcano** [comment]
Hi @valentinpalkovic, thank you for bringing Floating UI into the table. I had not considered it, but I have now taken a good look at it.

As for Radix UI and React Aria, being fully honest, I had gone through the docs of both of them, but had not tested them out to see which API was easier to work with, or felt more friendly and flexible. So I decided to prepare a small [codesandbox](https://codesandbox.io/s/storybook-dropdowns-k360sr?file=/src/RadixUIDropdown/index.tsx) testing both Radix UI and React Aria, which has actually changed my mind on both libraries.

Radix UI felt a bit nicer to work with. For some reason, while going through its docs, I thought it was less flexible in terms of styling, but it actually allow us to "add styles with our preferred CSS technology", as noted by their documentation.

As for React Aria, it's evidenced in the [codesandbox](https://codesandbox.io/s/storybook-dropdowns-k360sr) that it requires a lot more "setup" and boilerplate in general than Radix UI, and it seems like the "dependency" between react-aria and react-stately would be a bit tricky to replicate with a different library (other than react-stately), or with custom code written from scratch, which seems a bit much.

I have also checked Floating UI, and will link a [codesandbox](https://codesandbox.io/s/admiring-lamport-5wt3yg?file=/src/DropdownMenu.tsx:3125-3182) taken from their [documentation](https://floating-ui.com/docs/react-examples).

It definitely looks better than popper.js, and has been built with accessibility in mind, although from the codesandbox we can tell that its integration requires a lot of manual work and boilerplate. Its API also looks more complex than React Aria's one, which makes me think it might not be a quick win after all, but again, it's all open to discussion, of course.

If you could check the codesandboxes and let me know what your thoughts are, that would be awesome!

**Comment by danielmarcano** [comment]
For the sake of consistency, here's the information of both Radix UI and Floating UI:

Radix UI:

- It's great and flexible. We can use their components as-is and style them however we want, or pass them our custom components as children of their components

- It has no opinion on styling, and allows us to use any CSS technology / flavor

- The unpacked size of [@radix-ui/react-dropdown-menu](https://www.npmjs.com/package/@radix-ui/react-dropdown-menu) is 145 kB. The library encourages ["incremental adoption"](https://www.radix-ui.com/docs/primitives/overview/introduction#incremental-adoption), so each component has its own package.

- It's actively maintained, and has over 420K weekly downloads

- It fully supports React 18, as mentioned in these [release notes](https://www.radix-ui.com/docs/primitives/overview/releases#july-21-2022)

- It fully supports TypeScript, all their components' packages have built-in TypeScript types

Floating UI:

- It's also quite flexible, although it requires some manual work, accessibility-wise

- Fully flexible, and allows us to use any CSS technology / flavor

- The unpacked size of [@floating-ui/react](https://www.npmjs.com/package/@floating-ui/react) is 568 kB

- It's actively maintained, and has over 450K weekly downloads

- It fully supports TypeScript, the package has built-in TypeScript types

- It seems like it fully supports React 18, according to the [release notes](https://github.com/floating-ui/floating-ui/releases?q=react+18&expanded=true)

**Comment by danielmarcano** [comment]
Hi there, @valentinpalkovic @JReinhold, hope you are doing great!

Have you been able to go through the latest messages / have some more thoughts?

**Comment by valentinpalkovic** [comment]
Hi @danielmarcano,

sorry for the late reply, but I was on vacation :)

Thank you very much for your detailed analysis. This is very helpful and should make decision-making easier. As far as I can tell, even Radix UI uses Popper.JS in the background.

I now have a question, however, that the @storybookjs/core maintainers will have to answer in a closed manner:

Do we now want to make only the dropdown elements accessible, or do we have as a goal to replace more components with Radix UI components? If we only want to make the dropdown elements accessible, integrating Floating UI would be easier since a) there is a migration guide from Popper.js to Floating UI, and b) we would probably not cause a breaking change by doing so. In the long run, it would probably make sense to rely on a library like Radix UI, since we can then handle the accessibility issue uniformly with one library.

Based on your analysis, I would personally lean towards using Radix UI. Would you be willing to work on this topic and create a PR with the needed refactorings, should the team decide to go with this solution?

**Reply by danielmarcano** [reply]
Hi @valentinpalkovic, thank you for the update.

Regarding Radix UI using Popper.JS, initially I thought the same. However, upon further investigation, I discovered that Radix UI has actually developed its own "popper" implementation from scratch. You can find the source code for their "react-popper" [here](https://github.com/radix-ui/primitives/tree/main/packages/react/popper).

In response to your question, I am definitely willing to take on this task and create a PR with the necessary updates, whether it involves Radix UI or Floating UI, based on the decision made by the @storybookjs/core team.

Please keep me informed about the outcome and conclusion of that discussion through this RFC. I look forward to working on this!

**Reply by ndelangen** [reply]
@danielmarcano thank you so much for your interest!

The storybook core team met up last Monday UTC evening on discord (open meeting, feel welcome to join!), and we discussed this.

A few arguments were made:
- The Popper component is part of `@storybook/components` which is a public API.
- Making changes that are backwards incompatible to public APIs must be done in a major release.

We could go down 2 possible paths, and I'd like to hear your input:
- We upgrade to `floating-ui` because that's an iteration of popper, and thus it's APIs are effectively the same? If the APIs are the same, we can release this in a minor version bump.
- We create an new component based on radix, with the exact same purpose. We deprecate the old component, write a migration guide, and remove the popper based component in `8.0`
- We do both

Long term we'd love to use Radix everywhere, if that helps us be more accessible. We estimate this will be a slow process.

Storybook 8.0 is scheduled to be released around the end of this year (2023).

**Reply by danielmarcano** [reply]
Hi, @ndelangen.

Thanks for the information!

I have a doubt regarding the fact that @storybook/components is a public API.

What are the use-cases for that API outside the scope of the Storybook.js project? Just asking to have more context on this.

As for our options, have you taken a look at the migration guide between Popper.js and Floating UI to see if they share the same API? Otherwise, I could do that this week and leave my findings in this RFC.

If they happened to have different APIs, both options (Floating UI or Radix UI) could only be done in a major release, right? In which case, it might be best to simply go ahead with Radix UI.

Otherwise, if they do have the same API, we could do both:

- Add Floating UI as soon as possible, in a minor bump.

- Add Radix UI drop-down component in Storybook's upcoming major release, by the end of the year.

What are your thoughts on this?

**Reply by ndelangen** [reply]
> What are the use-cases for that API outside the scope of the Storybook.js project?

Addons. they often inject new UI into the manager, and use our components to do it.
That's the purpose of having `@storybook/components` otherwise we'd just keep all the components making up the storybook UI in `@storybook/manager` (which is NOT a public API).

> have you taken a look at the migration guide between Popper.js and Floating UI 
I have not personally, no. @valentinpalkovic mentioned this in the meeting, I believe.

> If they happened to have different APIs, both options (Floating UI or Radix UI) could only be done in a major release, right? In which case, it might be best to simply go ahead with Radix UI.

Yes, if both have incompatible changing in their interfaces, we might as well choose Radix; it's the better option of the 2.

We could also investigate it the existing public API can be crafted on top of re-implementation based on Radix.
If this can be done, the breaking change disappears and we can release it in a minor version bump.

**Reply by danielmarcano** [reply]
Thank you, @ndelangen. That's quite helpful to know!

Regarding the best approach or way to proceed, here's my suggestion:

In the coming week, I can review the APIs of both Popper.js and Floating UI to identify any potential breaking changes between the two. I will share my findings with you here.

If the APIs are similar, and migrating from Popper.js to Floating UI doesn't cause any issues or break the existing functionality, I would begin working on a pull request (PR) to migrate from Popper.js to Floating UI.

However, if there are significant differences or if the migration would likely break things, an alternative approach would be to begin working on a PR with Radix UI. In this case, I could evaluate if the existing logic can remain intact and integrate with Radix UI without causing any disruptions. Otherwise, this Radix UI contribution could be a part of the upcoming major version upgrade of Storybook.

**Comment by JesusTheHun** [comment]
There is [Headless UI](https://headlessui.com) and [Headless UI Float](https://headlessui-float.vercel.app).

The former is from the authors of TailwindCSS and the later is a community made thin wrapper that uses [Floating UI](https://floating-ui.com) under the hood. 
The are both written in TypeScript and support React` and Vue.

After migrating two projects, I can tell that they barely cost anything to adopt as long as you are not trying to over engineer your components, like creating a component that is too generic.

Headless UI is opinionated in term of accessibility though, and the programatic support is this in progress (calling `close()` is not always possible).
The lib works quite well, but there are some core issues that arise from version to version, so I strongly recommend that you pin the version that works for your needs. That is my only concern about this lib.
IRCC they are working on a v2 that is an internal rebuild, hopefully fixing those issues for good.

**Reply by danielmarcano** [reply]
Thank you for these suggestions, @JesusTheHun!

Could you please elaborate a bit more on the benefits of using Headless UI / Headless UI Float over Floating UI?

I have a feeling that both of them use Tailwind CSS, which I'm not sure would be desirable.

**Reply by JesusTheHun** [reply]
@danielmarcano Headless UI is a headless library: it does not come with any design of any form. They give you components to which you can add children and classes to create somethings that looks like you want. You can use TailwindCSS if you want, but it's not required at all. There is no design choice, that's why it's so convenient. Headless UI only brings the business logic and the accessibility.

Floating UI is a low level library to position floating element.
Headless UI Float is a thin community wrapper around Headless UI and Floating UI that just saves you time by doing all the boilerplate for you. It is also headless.

**Reply by JesusTheHun** [reply]
@danielmarcano honestly its so fast I can make you a PR if you want.

**Reply by danielmarcano** [reply]
@JesusTheHun since we had already decided that we could:

- Use Floating UI for the short-term, only if it has the same API as Popper.js

- Use Radix UI for the long-term, and introduce it in Storybook's next release

We would need to know if you are thinking of Headless UI for the short-term or the long-term option, and what its advantages would be over Radix UI and Floating UI.

What you have mentioned about the current versioning and the possible issues does not sound ideal, even though it looks like a nice library.

Also, if you could provide in this RFC the following details of Headless UI, so as to keep consistency, that would be truly valuable. Some of these you have already answered, but if you could document them in a list, like we have done with the other libraries' information, it would facilitate things:

- Ease of use of the API

- How flexible they are in terms of re-styling

- Size of the packages

- Status / whether they are actively maintained

- React 18 support

- TypeScript support

- Amount of components they have

**Reply by JesusTheHun** [reply]
@danielmarcano before anything I want to say that I do not own any share in TailwindLabs 😂

I'm thinking long term option.

The API of Floating UI is different than Popper.js' and has a lot of boilerplate.
But, to my eyes and in the eyes of many, it is the new standard.

Headless UI Float uses Floating UI under the hood. It just eliminates all the boilerplate for you.

The initial goal of this RFC was to replace dropdowns to make them more accessible. 
Is the goal still the same or the scope has extended?

Now, I will answers your questions with the usage of both Headless UI and Headless UI Float : 

- [x] Ease of use of the API: Very easy. Easier than Radix UI.
- [x] How flexible they are in terms of re-styling: Complete freedom, from the html template to the CSS. You can use the styled components you are currently using.
- [x] Size of the packages: Quite big since it's a single package (vs Radix that has a package per component), but it supports tree shaking so I believe it is irrelevant isn't it? 
- [x] Status / whether they are actively maintained: Actively maintained, with a company behind it.
- [x] React 18 support
- [x] TypeScript support: built in Typescript 
- [x] Amount of components they have: 9 components. That's why I asked if this RFC is still about dropdowns only?

**Reply by danielmarcano** [reply]
Hi, @JesusTheHun.

Sorry for the late reply, I was on vacation.

Thanks a bunch for all the information you have provided!

I asked whether you considered Headless UI to be a short-term or a long-term solution, since, in case the API between Popper.js and Floating UI was the same, we thought of adding a short-term solution, that would then be replaced by the long-term one, which is basically Radix UI.

Since the API of Headless UI is a bit unstable, as mentioned by your comments, I don't think it would be ideal for the long-term solution just now.

Also, as you have helped us understand that Popper.js and Floating UI have different APIs, I would say we will incline for the long-term solution with Radix UI, whose PR I will take care of.

@valentinpalkovic @JReinhold does this sound good to you both?

If that's so, I would begin working on this and preparing it for the upcoming major release of Storybook, which is due mid or late December, if I understood correctly.

**Reply by valentinpalkovic** [reply]
Hi @danielmarcano 

I am also in favor of going with the long-term solution, as the assumption of having the same API's in popper.js and Floating UI is wrong.

You have green lights from my side! @JReinhold, @ndelangen Any additions?

**Comment by JReinhold** [comment]
Sorry for being away from this for a long time, thanks for your hard work and research @danielmarcano.

I agree with @valentinpalkovic, the conclusion to this RFC from the core team:

**Introduce Radix UI to the components.**

- Short-term replace the menu dropdown components with ones based on Radix UI
- We're starting work on a new experimental UI package that will incrementally replace `@storybook/components`, this work would fit perfectly into that project. (first draft [here](https://github.com/storybookjs/storybook/pull/23356))
- Long-term introduce Radix UI more throughout the components to improve the overall accessibility and usage of the components.
- @danielmarcano you should sync up with @cdedreuille to get started on the work in the new experimental UI. You can make great things together, I might join.

🚀

cc @vanessayuenn

**Reply by danielmarcano** [reply]
Thanks for your kind comments and for outlining the next steps, @JReinhold!

As for this RFC per se, I just have two doubts:

- Do we close the discussion?

- How do we document the completion of this RFC once we begin working on the Radix UI drop-down as part of the new components you have mentioned?

**Comment by cdedreuille** [comment]
Hi @danielmarcano 👋,

I'm just catching up on your RFC. Fantastic job. Thanks a lot 🙏. We are in the process of improving our UI components for the exact reasons you mentioned but also to make sure it's coherent with the design we would like to implement. I'm part of the design team and I'm working on a new UI library that will allow us to shape an entirely new set of UI components for Storybook. I would love to share it with you and hopefully we can collaborate together on it if you're interested.

I'm writing a new RFC for this new library. As soon as it's ready I'll cross share here to get your eyes on it.

**Reply by danielmarcano** [reply]
Hi, @cdedreuille!

Thanks for your awesome feedback.

The new UI library that you are designing sounds great!

I would be happy to contribute with its development!

Let me know what you would need from my side to begin collaborating and working on this.

Perhaps chatting through a Discord channel in the Slack Discord server would facilitate things, for example?

**Reply by cdedreuille** [reply]
Feel free to DM me on discord (cdedreuille). I could show you around.

**Reply by danielmarcano** [reply]
@cdedreuille will do! Thank you

**Comment by cdedreuille** [comment]
I've been exploring Radix since we are using it in other places like on our Marketing site. Here's a draft PR for new form components I'm working on - https://github.com/storybookjs/storybook/pull/23469. I used Radix on the new `Select` component.

This component is heavily inspired from the Select component on Shadcn (https://ui.shadcn.com/docs/components/select).

I need to test it a bit more but so far it is working pretty well with all accessibility in mind.

**Comment by danielmarcano** [comment]
Hi everyone! I have just had an onboarding with @cdedreuille, and will begin working on the creation of the DropdownMenu for the new experimental UI package. Such component would be the beginning of this RFC's outcome.

Also, could we update the status of this RFC and leave it as "accepted" (I'm guessing this is done with a label), as mentioned in the [docs](https://storybook.js.org/docs/react/contribute/RFC)?
